### DISCUSSION TITLE
`pull-into`

### DISCUSSION THREAD

**Discussion by dabrahams** [magit/magit#4943] [Ideas]
I'm constantly updating my local main branch to stay in sync with upstream, but the fact that I have to switch off the current branch to do that causes churn in my working copy, which causes more to be rebuilt than necessary.  It would be great to have a `pull-into` option that lets me specify the branch into which changes will be pulled, without having to switch to that branch.

The most obvious implementation uses worktrees, and if the branch was already checked out in a worktree, I wouldn't mind if the worktree was updated.

Thanks Jonas, for all your great work!

**Comment by tarsius** [comment]
Occasionally I wish such a command already existed myself. I'll probably implement it eventually.

I like to use `master` as the upstream of `new-feature` and you seem to do the same. There are certainly benefits to that approach; but this problem is at least one downsides. The alternative is to use `origin/master` as the upstream instead of `master`. Unlike `refs/heads/master`, `refs/remotes/origin/master` *is* automatically updated when we `git pull`, so in a sense the "update the upstream without having to check it out first" feature already exists, we just both have opted to not use it.

**Reply by dabrahams** [reply]
Not sure if that would work the same way. I am often pulling into main so I can spin a new branch off the tip. There's not much point in having a main branch around otherwise

**Reply by tarsius** [reply]
> Not sure if that would work the same way. I am often pulling into main so I can spin a new branch off the tip.

If you use `origin/main` as the upstream of your feature branches, then `f u` updates the upstream branch (and all other `refs/remotes/origin/*`). So there is no need to update `main` (or any other branches) individually, unless of course you actually want to work on that branch, in which case you have to check it out anyway.

Of course there might be reasons why you want to use `main` as the upstream, not `origin/main`. For example to reduce the risk of accidentally pushing to `origin/main`, or because you sometimes *want* to lag behind `origin/main`.

If you choose to use `origin/main` as the upstream, then you should set `magit-prefer-remote-upstream` to `t`.

> There's not much point in having a main branch around otherwise

If you never work on that directly and always use `origin/main` (or another remote branch) as the upstream of your feature branches, then you can of course delete `main`. I am not sure I understood what you meant by that, so maybe this response isn't appropriate.

**Reply by tarsius** [reply]
Given that I have just rejected two other proposals by you, I would like to mention, that this has a change (but no guarantee) of being implemented by me.

**Reply by tarsius** [reply]
If added `magit-pull-into-upstream` in 05d7ea9870fead1f93def86bc1935dd48e4c77ee.

It's disabled in the pull menu by default, see https://docs.magit.vc/transient/Enabling-and-Disabling-Suffixes.html.
